src/crystal/system/win32/event_loop_iocp.cr

```diff
@@ -1,14 +1,30 @@
+require "c/iocp"
 require "crystal/system/print_error"
 
 module Crystal::EventLoop
-  @@queue = Deque(Fiber).new
+  @@queue = Deque(Crystal::Event).new
 
   # Runs the event loop.
   def self.run_once : Nil
-    next_fiber = @@queue.pop?
+    unless @@queue.empty?
+      next_event = @@queue.min_by { |e| e.wake_in }
```

Comment:
Agreed, would a priority queue be a better solution here?

Reply:
Maybe a dequeue is actually fine. You could sort at insertion and always pop the first entry.

Reply:
Well. There are three operations:
a: Insertion.
b: pop event of least timestamp
c: remove event matching what the completion returned. 

With unsorted deque b and c is slow (O(n)). With sorted deque a and c is slow (same). There are alternatives out there ( https://brilliant.org/wiki/pairing-heap/#summary-of-the-running-times-of-pairing-heaps for example, but there are probably even better examples out there) that are **way** faster. Also it might be a good idea to separate the queue into sleepers and completions - the completions could be served by a hash table and then the timeouts would only need a and b, which may allow for even better algorithms.

But as I said, it is hard enough to get things right, so it would probably be better to focus on that - improving this is quite simple and restricted once other things are out of the way. Getting the program flow running with some io and sleep operations is a lot more at this stage.

BTW, is the intention to support threads in the first iteration of this, or is single thread ok? If multithread is part of this iteration then the queue would likely need to be synchronized?

Reply:
Thanks for breaking down and highlighting the issues with the deque structure here. It neat to see Big O put into context with a real world example rather than just staring at an equation on a whiteboard in class. 

I had considered a solution with separate queues based on the event type, but it became too complicated. Instead I think the pairing heap will work best by combining efficiency and relative simplicity in implementation. [I'm working on a priority queue based on a min pairing heap, though right now it's just a regular min heap.](https://gist.github.com/incognitorobito/7dd167c4588b93941f83bedb954a2490) In order to preserve the O(log(n)) runtime for arbitrary deletes, I store a hash table mapping the objects to their locations in the heap. That way when deleting an arbitrary element the lookup is O(1) and then the heapify is O(log(n)). It will take me a little longer to put the min pairing heap together since work is now picking up but I want to do my best here.

Reply:
Alternate implementation (of the heap): https://github.com/yxhuvud/pairing_heap/blob/master/src/pairing_heap.cr . Feel free to pick what you want.

However, remember that it is probably good to focus on getting the event loop right as it allows up completing work in other parts of the io handling that is blocked by the event loop now. This can be improved after merge of the rest of the loop.